lnrpc/routerrpc/router_server.go

```diff
@@ -401,11 +447,276 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
+// hints don't indicate an LSP, they are passed as arguments to the
+// SendPayment_V2 method, which enable it to send probe payments to the payment
+// request destination.
+func (s *Server) probeInvoice(ctx context.Context, paymentRequest string,
+	timeout int32) (*RouteFeeResponse, error) {
+
+	payReq, err := zpay32.Decode(
+		paymentRequest, s.cfg.RouterBackend.ActiveNetParams,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate random payment hash, so we can be sure that the target of
+	// the probe payment doesn't have the preimage to settle the htlc.
+	var paymentHash lntypes.Hash
+	_, err = crand.Read(paymentHash[:])
+	if err != nil {
+		return nil, fmt.Errorf("cannot generate random probe "+
+			"preimage: %w", err)
+	}
+
+	probeRequest := &SendPaymentRequest{
+		TimeoutSeconds:   timeout,
+		Dest:             payReq.Destination.SerializeCompressed(),
+		MaxParts:         1,
+		AllowSelfPayment: false,
+		AmtMsat:          int64(*payReq.MilliSat),
+		PaymentHash:      paymentHash[:],
+		FeeLimitSat:      100_000_000,
+		PaymentAddr:      payReq.PaymentAddr[:],
+	}
+
+	hints := payReq.RouteHints
+	isLsp, lspAdjustedRouteHints, lspHopHint := isLSP(hints)
+	if !isLsp {
+		probeRequest.RouteHints = invoicesrpc.CreateRPCRouteHints(hints)
+		return s.sendProbePayment(ctx, probeRequest)
+	}
+
+	if len(lspAdjustedRouteHints) == 0 && lspHopHint == nil {
+		return nil, errors.New("invalid lsp route hints")
+	}
+
+	// The adjusted route hints serve the payment probe to find the last
+	// public hop to the LSP on the route.
+	probeRequest.Dest = lspHopHint.NodeID.SerializeCompressed()
+	probeRequest.RouteHints = invoicesrpc.CreateRPCRouteHints(
+		lspAdjustedRouteHints,
+	)
+
+	// The payment probe will be able to calculate the fee up until the LSP
+	// node. The fee of the last hop has to be calculated manually. Since
+	// the last hop's fee amount has to be sent across the payment path we
+	// have to add it to the original payment amount. Only then will the
+	// payment probe be able to determine the correct fee to the last hop
+	// prior to the private destination. For example, if the user wants to
+	// send 1000 sats to a private destination and the last hop's fee is 10
+	// sats, then 1010 sats will have to arrive at the last hop. This means
+	// that the probe has to be dispatched with 1010 sats to correctly
+	// calculate the routing fee.
+	//
+	// Calculate the hop fee for the last hop manually.
+	hopFee := lspHopHint.HopFee(int64(*payReq.MilliSat))
+
+	// Add the last hop's fee to the requested payment amount that we want
+	// to get an estimate for.
+	probeRequest.AmtMsat += hopFee
+
+	// Dispatch the payment probe with adjusted fee amount.
+	resp, err := s.sendProbePayment(ctx, probeRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	// The probe succeeded, so we can add the last hop's fee to the
+	// to fee the payment probe returned.
+	resp.RoutingFeeMsat += hopFee
```

Comment:
Won't the returned `TimeLockDelay` be innacurate here? It should take into account the delay set in `lspHopHint`.

Reply:
nice catch !